server.js

```diff
@@ -29,20 +30,47 @@ app.get('/static', (req, res) => {
     res.render('pages/static')
 })
 
+const stateStore = new Map()
+
 // Final plugin, renders dynamic HTML
 app.get('/dynamic', async (req, res) => {
-    // Get the authorization code from the URL parameters
-    const auth_code = req.query.code
-
-    const api_keys = `${config.api.client_id}:${config.api.client_secret}`
-    const base64_keys = Buffer.from(api_keys).toString('base64')
-
     // The REDIRECT_URI must be added to the institution settings in Banno People
     // *NOTE* it is case sensitive
     const REDIRECT_URI = `http://localhost:${config.app_port}/dynamic`
 
+    let state
+    let codeVerifier
+    if (!req.query.code || !req.query.state) {
+        if (req.query.state) {
+            stateStore.delete(req.query.state)
+        }
+
+        // create the PKCE code verifier and stash it for later
+        codeVerifier = crypto.randomBytes(60).toString('hex').slice(0, 128)
+        const CODE_CHALLENGE = crypto.createHash('sha256')
+            .update(Buffer.from(codeVerifier)).digest('base64')
+            .replace(/=/g, '')
+            .replace(/\+/g, '-')
+            .replace(/\//g, '_')
+
+        state = crypto.randomBytes(60).toString('hex').slice(0, 128)
+        stateStore.set(state, codeVerifier)
+        res.cookie(`STATE_${state}`, codeVerifier, {httpOnly: true, sameSite: 'lax'})
+
+        // redirect
+        res.redirect(`${config.api.environment}/a/consumer/api/v0/oidc/auth?scope=openid&response_type=code&client_id=${config.api.client_id}&redirect_uri=${encodeURIComponent(REDIRECT_URI)}&state=${state}&code_challenge=${CODE_CHALLENGE}&code_challenge_method=S256`)
```

Comment:
**future:** This whole chunk probably needs some clarifying comments as it is a bit non-obvious what this is doing here unless the reader is very knowledgeable about OAuth+OpenID Connect in general and PKCE in particular.

We might also want a link to something like the Authentication (Command Line) Quickstart, although that's just a quick thought. https://jackhenry.dev/open-api-docs/consumer-api/quickstarts/AuthenticationCommandLine/